README.md
# ExInsights

[![Hex.pm](https://img.shields.io/hexpm/v/ex_insights.svg?style=flat-square&colorB=6e347e)](https://hex.pm/packages/ex_insights)

Elixir client library to log telemetry data on Azure Application Insights.

## Installation

Install from hex by adding `ex_insights` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:ex_insights, "~> 0.8"}
  ]
end
```

You then need to start the `ExInsights.Supervisor` in your supervision tree. Example:

```elixir
# inside the init/1 callback of a supervisor in your supervision tree (or start/2 of application.ex)

children = [
  # ...
  {ExInsights.Supervisor, instrumentation_key: "0000-1111-2222-3333"}
]

Supervisor.init(children, strategy: :one_for_one)
```
### Note
For migrating to 0.8.x and later versions read the migration instructions below

## Configuration
ExInsights supports the following configuration options you can set when starting the `ExInsights.Supervisor`

* `:instrumentation_key`: You can set the instrumentation key (Azure App Insights secret key) to use for sending data to Azure. It can be overriden on each `ExInsights.track_xxx` request. Needs to be present either by setting it here or on every request, otherwise the code will raise with an error (string)
* `:flush_inteval_secs`: the number of seconds every which the client will send the telemetry data to Azure. Default is every 30 seconds since the last flushing (non-negative integer)
* `:client_module`: the module to use for actually sending the requests. Mostly useful in tests (atom)

Example:

```elixir
options = [
  instrumentation_key: "0000-1111-2222-3333",
  flush_interval_secs: 30
]

children = [
  # ...
  {ExInsights.Supervisor, options}
]

Supervisor.init(children, strategy: :one_for_one)
```

## Usage

### Basic Usage

All public tracking methods are under the `ExInsights` module. Examples:

```elixir
# will post a click custom_event to azure
ExInsights.track_event("click")

# with custom defined property "type" and measurement "count"
ExInsights.track_event("click", %{type: "button"}, %{count: 2})

# send custom metric data. Does not support aggregated data (count/stdDev, min, max)
ExInsights.track_metric("bananas", 10)

# log arbitrary data
ExInsights.track_trace("1-2-3 boom", :warning)

# log time taken for requests to external resources, eg. database or http service calls
ExInsights.track_dependency("get_user_balance", "http://my.api/get_balance/aviator1", 1500, true, "user", "my.api")

# log telemetry data about the incoming request processed by the application
ExInsights.track_request("homepage", "http://my.site.com/", "HomePageComponent", 85, 200, true)
```

For more details and optional arguments look at the [`ExInsights`](https://hexdocs.pm/ex_insights/ExInsights.html) module documentation.

### Alternative usage

You can also decorate methods you need to track using [decorators](https://github.com/arjan/decorator).

```elixir
# Make sure to add the following line before using any decorators
use ExInsights.Decoration.Attributes

# add the @decorate track_xxx() attribute right above each function you need to track

@decorate track_event() # will log the "update_user_email" event in AppInsights on function entry
def update_user_email(email, user) do
  # ...
end

@decorate track_dependency("user-actions") # put under dependency type:user-actions in AppInsights UI
def login_user(user) do
  # ... maybe call external api here
end

@decorate track_exception() # will track errors and exits
def dangerous_stuff do
  # ... do work that may fail
end
```

## Migrating from 0.7.x to 0.8.x
In order to make the library configuration more flexible and in accordance to [library development guidelines for configuration](https://hexdocs.pm/elixir/master/library-guidelines.html#avoid-application-configuration), support for configuring the client by setting options directly inside config.exs files was dropped and configuration now needs to happen on supervisor startup.

For example instead of globally configuring ex_insights inside config.exs,
```elixir
# No longer supported
config :ex_insights,
  instrumentation_key: "0000-1111-2222-3333"
```

you now need to pass this option to the `ExInsights.Supervisor` directly instead
```elixir
children = [
  ...
  {ExInsights.Supervisor, instrumentation_key: "0000-1111-2222-3333"}
]
```

If you need to keep reading this value from your config files you can still do something like this
```elixir
# in config.exs
config :my_app, ex_insights_instrumentation_key: "0000-1111-2222-3333"

# in you supervisor.ex file
def read_instrumentation_key_from_config do
  Application.get_env(:my_app, :ex_insights_instrumentation_key)
end

def init(_) do
  ...
  children = [
    # ...
    {ExInsights.Supervisor, instrumentation_key: read_instrumentation_key_from_config()}
  ]
end
```

## Inner workings
* Calling any tracking function `ExInsights.track_xxx` from your code will not immediately send the data to Azure. It will instead be aggregated in memory until the `flush_timer` is triggered (every 30 secs, configurable) and the data will be batch sent.
* When the application shuts down it will attempt to flush any remaining data.
* If you are behind a firewall (usually happens in production deployments) make sure your network rules **allow HTTP POSTs to https://dc.services.visualstudio.com**
* If requests to azure tracking services fail (network or server errors or bad requests) you will not be alerted.
* `track_dependency` and `track_exception` decorators will try to `rescue`/`catch` any errors (and log those) and then re-raise the error / exit as appropriate. This is a different (but hopefully working) approach than what the AppSignal guys do (a separate process monitoring crashes)


CHANGELOG.md
# Changelog

## 0.8.1
- Update deps
## 0.8.0
- __Breaking change__: ExInsights is no longer an application, you now need to start the Supervisor manually
- __Breaking change__: Configuring from `config.exs` was dropped in favor of passing arguments to the Supervisor. This change along with the one above should help making the library more flexible to configure
- _Possibly breaking change_: replaced Poison with Jason as the default json encoder
- cleaned up internals to make extending the library easier. Public interface has not changed.

## 0.7.0
- __Breaking change__: start time and dependency id parameters as part of request and track dependency payload (thanks @huderlem)

## 0.6.0
- __Breaking change__: custom tags as part of payload (thanks @huderlem)

## 0.5.2
- allow setting id explicitly on track_request

## 0.5.1
- fix dialyzer issue

## 0.5.0
- allow setting instrumentation key on a per request basis

## 0.4.1
- added measurements to request tracking (thanks @DevExpDev)

## 0.4.0
- added request tracking (thanks @lafka)
- formatted code files
- updated dependencies

## 0.3.1
- Minor changes in test suite
- Updated typespecs

## 0.3.0
- Added `track_exception`
- Added decorators for `track_event`, `track_exception`, and `track_dependency`
- Updated docs
- Cleaned up specs and added dialyzer

## 0.2.0
- Added `track_dependency` and `track_trace`
- Added `flush_interval_secs` config option
- Documentation changes

## 0.1.0
- Initial version
- Support for `track_event` and `track_metric`


lib/data/envelope.ex
defmodule ExInsights.Data.Envelope do
  @moduledoc ~S"""
  Track request envelope

  Envelope data looks like this
  ```json
  {
    "time": "2017-08-24T08:55:56.968Z",
    "iKey": "some-guid-value-key",
    "name": "Microsoft.ApplicationInsights.someguidvaluekey.Event",
    "tags": {
      "ai.session.id": "SLzGH",
      "ai.device.id": "browser",
      "ai.device.type": "Browser",
      "ai.internal.sdkVersion": "javascript:1.0.11",
      "ai.user.id": "V2Yph",
      "ai.operation.id": "VKgP+",
      "ai.operation.name": "/"
    },
    "data": {
      "baseType": "EventData",
      "baseData": {
        "ver": 2,
        "name": "button clicked",
        "properties": {
          "click type": "double click"
        },
        "measurements": {
          "clicks": 2
        }
      }
    }
  }
  ```
  """

  @data_version 2
  @app_version Mix.Project.config()[:version]

  defstruct [
    :time,
    :name,
    :iKey,
    :tags,
    :data
  ]

  @type t :: %__MODULE__{
    time: String.t(),
    name: String.t() | nil,
    iKey: String.t() | nil,
    tags: map(),
    data: map()
  }

  @doc """
  Creates a new envelope for sending a single tracked item to app insights. Intended for internal use only.
  """
  def create(%{} = data, type, %DateTime{} = time, %{} = tags) when is_binary(type) do
    %__MODULE__{
      time: DateTime.to_iso8601(time),
      tags: tags,
      data: %{
        baseType: "#{type}Data",
        baseData: Map.put(data, :ver, @data_version)
      }
    }
  end

  @doc """
  Provides common tags for all track requests. Intended for internal use only.
  """
  def get_tags() do
    %{
      "ai.internal.sdkVersion": "elixir:#{@app_version}"
    }
  end

  def set_instrumentation_key(%__MODULE__{data: %{baseType: baseType}} = envelope, instrumentation_key) do
    type = String.replace(baseType, "Data", "")
    name = "Microsoft.ApplicationInsights.#{String.replace(instrumentation_key, "-", "")}.#{type}"
    %{envelope | name: name, iKey: instrumentation_key}
  end

  def ensure_instrumentation_key_present(%__MODULE__{iKey: key}) when key in [nil, ""],
  do:
    raise("""
    Azure app insights instrumentation key not set!
    1) First get your key as described in the docs https://docs.microsoft.com/en-us/azure/application-insights/app-insights-cloudservices
    2) Then set it either
      a) during application execution using Application.put_env(:ex_insights, :instrumentation_key, "0000-1111-2222-3333"), OR
      b) in your config.exs file using either the vanilla or {:system, "KEY"} syntax. Examples:

        config :ex_insights,
          instrumentation_key: "00000-11111-2222-33333"

        OR

        config :ex_insights,
          instrumentation_key: {:system, "INSTRUMENTATION_KEY"}

        When using the {:system, "KEY"} syntax make sure that the env variable is defined on system startup, ie to start your app you should do
        INSTRUMENTATION_KEY=0000-1111-2222-333 iex -S mix OR
      c) as a paremeter along with each request, ie: ExInsights.track_event(..., instrumentation_key)
    """)

    def ensure_instrumentation_key_present(%__MODULE__{} = envelope), do: envelope
end
